executor/src/witgen/jit/processor.rs

```diff
@@ -483,6 +583,59 @@ fn machine_call_params<T: FieldElement>(
     })
 }
 
+/// Returns all unknown but relevant (e.g. not multiplied by a value known to be zero)
+/// variables in the expression evaluated on the given row offset.
+fn unknown_relevant_variables<T: FieldElement, FixedEval: FixedEvaluator<T>>(
+    expr: &AlgebraicExpression<T>,
+    witgen: &WitgenInference<'_, T, FixedEval>,
+    row_offset: i32,
+) -> HashSet<Variable> {
+    match expr {
+        AlgebraicExpression::Reference(algebraic_reference) => {
+            let var = Variable::from_reference(algebraic_reference, row_offset);
+            (!witgen.is_known(&var))
+                .then_some(var)
+                .into_iter()
+                .collect()
+        }
+        // TODO once we support them, we should turn them into the proper variables here.
+        AlgebraicExpression::PublicReference(_) | AlgebraicExpression::Challenge(_) => {
+            Default::default()
+        }
+        AlgebraicExpression::Number(_) => Default::default(),
+        AlgebraicExpression::BinaryOperation(AlgebraicBinaryOperation { left, op, right }) => {
+            if *op == AlgebraicBinaryOperator::Mul {
+                let left_val = try_to_evaluate_to_known_number(left.as_ref(), witgen, row_offset);
+                let right_val = try_to_evaluate_to_known_number(right.as_ref(), witgen, row_offset);
+                if left_val == Some(T::from(0)) || right_val == Some(T::from(0)) {
+                    return Default::default();
+                }
+            }
+            let mut vars = unknown_relevant_variables(left.as_ref(), witgen, row_offset);
+            vars.extend(unknown_relevant_variables(
+                right.as_ref(),
+                witgen,
+                row_offset,
+            ));
+            vars
```

Comment:
```suggestion
            unknown_relevant_variables(left.as_ref(), witgen, row_offset)
                .into_iter()
                .chain(unknown_relevant_variables(right.as_ref(), witgen, row_offset).into_iter())
                .collect()
```

Reply:
This needs to re-hash the left variables and the right variables. Maybe it would be best to return an iterator and just collect it into a hash set at the very end...

Reply:
Wait, are you saying that if you extend a hash set by a hash set, it does not re-hash the items? I'd be surprised, because the signature is `fn extend<I: IntoIterator<Item = T>>(&mut self, iter: I)` AFAICS and also the number of buckets might be different, no?